utils/resource-monitor/graph_pidstat.py

```diff
@@ -0,0 +1,215 @@
+
+    Args:
+      num_proc: Number of processes the batch job is set to run with
+      data_description: Description of the dataset
+      output_parquet_path: Directory of the output parquet files
+      output_results_path: Directory of the output csv files and graphs
+
+    Returns:
+      None
+    """
+    if not os.path.exists(output_results_path + str(data_description) + '/graphs'): os.makedirs(output_results_path + str(data_description) + '/graphs')
+    if not os.path.exists(output_results_path + str(data_description) + '/tables'): os.makedirs(output_results_path + str(data_description) + '/tables')
+    if not os.path.exists(output_results_path + str(data_description) + '/raw'): os.makedirs(output_results_path + str(data_description) + '/raw')
+    subprocess.call(['sh', './monitor_pipeline.sh', num_proc, data_description, output_parquet_path, output_results_path])
+
+def make_pidstat_tables(file_name, type):
+    """
+    Creates pandas tables from raw pidstat data.
+
+    Args:
+      file_name: Name of the .txt file containing raw pidstat output
+      type: Type of the resource relevant to the table (Pipeline, Server, or DB)
+
+    Returns:
+      Pandas table with CPU, memory and I/O usage info for the resource.
+    """
+
+    #Parse the raw .txt file into a pandas dataframe
+    total_usage = []
+    seconds_elapsed = 0
+    with open(file_name) as f:
+        for line in f:
+            usage = []
+            if line[0].isdigit():
+                usage = line.split()
+
+                if type == 'DB ' and usage[-3] != 'hapi':
+                    continue
+
+                del usage[:8]
+                del usage[9:]
+                del usage[1:6]
+                usage.insert(0, seconds_elapsed)
+                total_usage.append(usage)
+            elif line[0] == '#':
+                seconds_elapsed += 1
+                
+    total_usage_df = pd.DataFrame(total_usage, columns = ['Seconds Elapsed', type + '%CPU', type + 'MEM', type + 'kBs Read / Second', type + 'kBs Written / Second'])
+    total_usage_df[type + 'MEM'] = total_usage_df[type + 'MEM'].apply(lambda x: float(x)/100*get_machine_mem())
+
+    #Cast types of all columns
+    for col in total_usage_df:
+        total_usage_df[col] = total_usage_df[col].astype(float)
+
+    #Summation by the second for the database table
+    if type == 'DB ':
+        total_usage_df['DB %CPU'] = total_usage_df.groupby(['Seconds Elapsed'])['DB %CPU'].transform(sum)
+        total_usage_df['DB MEM'] = total_usage_df.groupby(['Seconds Elapsed'])['DB MEM'].transform(sum)
+        total_usage_df['DB kBs Written / Second'] = total_usage_df.groupby(['Seconds Elapsed'])['DB kBs Written / Second'].transform(sum)
+        total_usage_df['DB kBs Read / Second'] = total_usage_df.groupby(['Seconds Elapsed'])['DB kBs Read / Second'].transform(sum)
+        total_usage_df = total_usage_df.drop_duplicates(subset=['Seconds Elapsed'])
+
+    total_usage_df[total_usage_df < 0] = 0
```

Comment:
Why is this needed?

Reply:
Right, but what I am wondering if why we may get a negative number in the `total_usage_df` at this point?

Reply:
Occasionally, the kB_rd/s and kB_wr/s fields from executing the pidstat command could display '-1', which I believe means the fields could not be generated correctly by pidstat. In those cases, I've configured the pandas data frame to treat '-1' as '0'.